File: Java/boj/Beginner/k_backtracking/boj_2580.java
package boj.Beginner.k_backtracking;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

public class boj_2580 {

    static StringTokenizer st;
    static final int MAX = 9;
    static final int[][] arr = new int[MAX][MAX];
    static final List<int[]> v = new ArrayList<>();
    static final BufferedReader br =
            new BufferedReader(new InputStreamReader(System.in));

    private static boolean chk(int row, int col, int i) {
        for (int k = 0; k < MAX; k++)
            if (arr[row][k] == i || arr[k][col] == i) return false;

        int vx = (row / 3) * 3;
        int vy = (col / 3) * 3;

        for (int a = vx; a < vx + 3; a++)
            for (int b = vy; b < vy + 3; b++)
                if (arr[a][b] == i) return false;

        return true;
    }

    private static boolean dfs(int cur) {
        if (cur == v.size()) return true;

        int row = v.get(cur)[0];
        int col = v.get(cur)[1];

        for (int i = 1; i <= MAX; i++) {
            if (chk(row, col, i)) {
                arr[row][col] = i;
                if (dfs(cur + 1)) return true;
                arr[row][col] = 0;
            }
        }

        return false;
    }

    public static void main(String[] args) throws IOException {
        for (int i = 0; i < MAX; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < MAX; j++) {
                arr[i][j] = Integer.parseInt(st.nextToken());
                if (arr[i][j] == 0) v.add(new int[] { i, j });
            }
        }

        dfs(0);

        for (var a : arr) {
            for (var b : a) System.out.print(b + " ");
            System.out.println();
        }
    }
}

File: Kotlin/src/main/kotlin/beginner/k_backtracking/boj_2580.kt
package beginner.k_backtracking

import java.util.StringTokenizer

private const val MAX = 9
private val arr = Array(MAX) { IntArray(MAX) }
private val v = ArrayList<Pair<Int, Int>>()

private fun chk(row: Int, col: Int, i: Int): Boolean {
    for (k in 0 until MAX)
        if (arr[row][k] == i || arr[k][col] == i) return false

    val vx = (row / 3) * 3
    val vy = (col / 3) * 3

    for (a in vx until vx + 3)
        for (b in vy until vy + 3)
            if (arr[a][b] == i) return false

    return true
}

private fun solve(cur: Int): Boolean {
    if (cur == v.size) return true

    val row = v[cur].first
    val col = v[cur].second

    for (i in 1..MAX) {
        if (chk(row, col, i)) {
            arr[row][col] = i
            if (solve(cur + 1)) return true
            arr[row][col] = 0
        }
    }

    return false
}

fun boj_2580() = with(System.`in`.bufferedReader()) {
    for (i in 0 until MAX) {
        val st = StringTokenizer(readLine())

        for (j in 0 until MAX) {
            arr[i][j] = st.nextToken().toInt()
            if (arr[i][j] == 0) v.add(Pair(i, j))
        }
    }

    solve(0)

    for (i in 0 until MAX) {
        for (j in 0 until MAX)
            print("${arr[i][j]} ")
        println()
    }
}